src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -192,13 +217,16 @@ private unsafe void Monitor(AsyncReadState state)
                         state.ThreadPoolBinding.FreeNativeOverlapped(overlappedPointer);
                     }
 
-                    // Clean up the thread pool binding created for the entire operation
-                    state.PreAllocatedOverlapped.Dispose();
-                    state.ThreadPoolBinding.Dispose();
+                    // Check whether the directory handle is still valid _before_ we dispose of the state,
+                    // which will invalidate the handle.
+                    bool wasValidDirectoryHandle = !IsHandleInvalid(state.DirectoryHandle);
+
+                    // Clean up the state created for the whole operation.
+                    state.Dispose();
 
                     // Finally, if the handle was for some reason changed or closed during this call,
                     // then don't throw an exception.  Otherwise, it's a valid error.
-                    if (!IsHandleInvalid(state.DirectoryHandle))
+                    if (wasValidDirectoryHandle)
                     {
                         OnError(new ErrorEventArgs(new Win32Exception()));
```

Comment:
This will do `Marshal.GetLastPInvokeError()` to capture the error returned by `ReadDirectoryChangesW`. A lot of code run in between that could have modified last error. We should capture the last error eight after `ReadDirectoryChangesW`.